### PR TITLE
Improved cw-2.py

### PR DESCRIPTION
**Pull request by Bubbler-4**
This PR introduces many improvements based on the discussions at #569 and [my own Kumite](https://www.codewars.com/kumite/5a724cec4a6b34f5d00000ba?sel=5a724cec4a6b34f5d00000ba).

---

# Testing Functions 

* Make testing functions non-blocking
* Change the expression inside `assert_not_equals` so that it can prevent operator injection hacks
* Provide `pass_`, `fail`, and `assert_approx_equals` with customizable precision (`pass` is a Python keyword)
* Replace hardcoded `print`s with `display`

# Describe and It

* Build the decorator version of `describe` and `it`, so the text fixture may look like this (and the decorator itself runs the code, so it doesn't need a separate runner function):

```python
@describe('describe-text')
def describe1():
    @it('it-text', before=f1, after=f2)
    def it1():
        # some test function
    @it('it-text')
    def it2():
        # some test function
```

* Properly close `describe` and `it` blocks in the test output
* Print the running times of `describe` and `it` blocks
* Provide `before` and `after` for `describe` and `it`

# Timeout Utility

* Apply a timeout limit on any block of code (the parameter is in seconds, and floats are accepted):

```python
@timeout(0.01)
def some_block():
    do_something()
    test_something()
```

* This is guaranteed to generate a failed test when the block timeouts

# The Unicode Error Trap

* Print using `"\uXXXX"` and `"\UXXXXXXXX"` literal style for tagged messages
* Print using `"&#XXXX;"` HTML escape style for simple logs
* This addresses #508 and #545, but not #558 (which, I guess, could be solvable by providing the source charset on the first line, which in turn requires some edits in the runner code, not `cw-2.py`)

### CODE REVIEW

**Comment by Bubbler-4** [comment]
@kazk Would it be OK to add runner-side changes upon this, if this PR isn't going to be merged soon?

**Comment by kazk** [comment]
@Bubbler-4 Thanks! Looks good to me so far :)
Can you elaborate on "runner-side" changes? If you mean [the fix to avoid concatenation](https://github.com/Codewars/codewars-runner-cli/issues/569#issuecomment-362130262), then I'm interested.

File: test/runners/python_spec.js [hunk 1]
```diff
@@ -225,3 +225,102 @@ describe('Output format commands', function() {
     });
   }
 });
+
+describe('Fixed and new features', function() {
```
Comment (kazk) [inline test/runners/python_spec.js]: Can you also add the test for the Unicode fix (#508, #545)?

Comment (kazk) [reply test/runners/python_spec.js]: > could be solvable by providing the source charset on the first line, which in turn requires some edits in the runner code

Do you mean adding `# coding: utf-8`?

Comment (Bubbler-4) [reply test/runners/python_spec.js]: Yes. But since we're going for multiple files approach, we'll need it in each file (at least preloaded, solution, and fixture).

Comment (kazk) [reply test/runners/python_spec.js]: If we go multiple files, can users just add them when needed?
This is currently impossible because we concatenate them and the magic comment is only valid if it's in the first or the second line.

Comment (Bubbler-4) [reply test/runners/python_spec.js]: Since we'll want to prepend something to the code when needed (just like in your `unittest` fix), I'd just add it to every file I create. At least it won't break existing things (because utf-8 is a superset of ascii).

**Comment by Bubbler-4** [comment]
For the runner changes, I'm thinking about these (blatantly copied from my kumite's comment section):

* Add to the first line `# -*- coding: utf-8 -*`- so we can use Unicode literals without `\u` escape
* Write each source file into `.py` files using `writeFileSync` and then import them in the runner code
* Remove `sanitizeStdErr` since its existence isn't relevant anymore, and giving full stack trace is much better choice
* Probably we can also write `runner.py` and run with `python runner.py` instead of `python -c ...` which is ugly af
* We don't even need to import `cw-2.py` to `runner.py`; what if we do so in `fixture.py`?

So it covers both [the discussion here](https://github.com/Codewars/codewars-runner-cli/issues/569#issuecomment-362130262) and #558, but it'll be really breaking changes.

**Comment by kazk** [comment]
We can't have any breaking changes for this repository. Which of those are breaking changes?

`# -*- coding: utf-8 -*-` and writing files looks fine to me. And removing `sanitizeStdErr` is technically a breaking change, but shouldn't have effects so I think we should be fine.


> - We don't even need to import `cw-2.py` to `runner.py`; what if we do so in `fixture.py`?

Yeah, I dislike how some languages runner do these. Also that `reload` and path manipulation shouldn't be necessary if we set `PYTHON_PATH` appropriately, right?

I think moving them to `fixture.py` is fine. My main concern with concatenation is line numbers in error messages not matching their solution.

**Comment by kazk** [comment]
Python 2.7 is printing `'\uac00\n'` when expected to include `'&#44032;'`
Python 3.4, 3.6 is throwing:
```
UnicodeEncodeError: 'ascii' codec can't encode character '\uac00' in position 0: ordinal not in range(128)
```

File: test/runners/python_spec.js [hunk 2]
```diff
@@ -322,5 +322,31 @@ describe('Fixed and new features', function() {
...
       });
     });
+
+    it(`should support unicode output (log) (Python${v} cw-2)`, function(done) {
+      runner.run({
+        language: 'python',
+        languageVersion: v,
+        testFramework: 'cw-2',
+        code: 'a = 1',
+        fixture: 'print("u\\uac00")',
```
Comment (kazk) [inline test/runners/python_spec.js]: Did you mean `fixture: 'print(u"\\uac00")',`? `"u` vs `u"`. Not sure if that makes any difference.

File: frameworks/python/cw-2.py [hunk 3]
```diff
@@ -1,24 +1,49 @@
 from __future__ import print_function
+import six
 
 class AssertException(Exception):
     pass
 
 
+'''Fix the dreaded Unicode Error Trap'''
+def print(*args, **kwargs):
```
Comment (kazk) [inline frameworks/python/cw-2.py]: This `print` doesn't seem to work when given non-string, e.g., `print(1)` throws
```
TypeError: 'int' object is not iterable
```
in Python 2.7 and
```
TypeError: print() argument after * must be a sequence, not map
```
in Python 3

File: frameworks/python/cw-2.py [hunk 4]
```diff
@@ -1,24 +1,49 @@
...
+    from sys import stdout
+    sep = kwargs.get('sep', ' ')
+    end = kwargs.get('end', '\n')
+    file = kwargs.get('file', stdout)
+    
+    def _replace(c):
+        if ord(c) >= 128: return u'&#{};'.format(ord(c))
+        return c
+    def _escape(s): return ''.join(_replace(c) for c in s)
```
Comment (kazk) [inline frameworks/python/cw-2.py]: In Python 2, the above error is from this `_escape`.

File: test/runners/python_spec.js (hunk 2 above)
Comment (kazk) [reply test/runners/python_spec.js]: For Python 2.7, fixing this should make the test pass.
```python
print(u"\uac00") #=> &#44032;
print("u\uac00") #=> u\uac00
```

**Comment by Bubbler-4** [comment]
I'm out today, so changes won't be made until tomorrow.

**Comment by kazk** [comment]
Don't worry, take your time. I just wanted to leave some feedback because I don't know if I have the time to review during the week.

**Comment by kazk** [comment]
@Bubbler-4 
For Python 3 throwing error, setting environment variable `PYTHONIOENCODING=UTF-8` might resolve the issue and it might not be necessary to escape them.

- https://stackoverflow.com/a/43989185
- https://chase-seibert.github.io/blog/2014/01/12/python-unicode-console-output.html

Setting environment variable can be done by changing
https://github.com/Codewars/codewars-runner-cli/blob/dbf943d62c8d6ed451d27ac1817c0f13b1de558c/lib/runners/python.js#L37

to

```javascript
    run({
      name: pythonCmd(opts),
      args: ['-c', code],
      options: {
        env: Object.assign({}, process.env, {PYTHONIOENCODING: 'UTF-8'})
      }
    });
```

**Comment by Bubbler-4** [comment]
Finally solved it, without using environment variables:

* Fall back to providing `uni_print` instead of overwriting `print` (I forgot Py2 runner doesn't have `from __future__ import print_function` at the beginning)
* Force cast to `six.text_type`, then apply some regex for better output in Py2

File: test/runners/python_spec.js [hunk 5]
```diff
@@ -329,9 +329,12 @@ describe('Fixed and new features', function() {
         languageVersion: v,
         testFramework: 'cw-2',
         code: 'a = 1',
-        fixture: 'print(1, "2", u"\\uac00")',
+        fixture: 'test.uni_print(1, "a", u"\\uac00", [314159, "b", u"\\uac01"])',
```
Comment (kazk) [inline test/runners/python_spec.js]: For Python 2, this is printing:
```
1 a &#44032; [314159, 'b', u'&#44033;']
```
image

https://travis-ci.org/Codewars/codewars-runner-cli/jobs/340822289#L1401

Comment (Bubbler-4) [reply test/runners/python_spec.js]: Should I remove the `u` flag in front of the literal? I think I'll leave it there, because `str` and `unicode` are two different things in Py2.

Comment (kazk) [reply test/runners/python_spec.js]: Ok, I didn't realize that was intentional. So it'll be rendered like this:
image

Comment (Bubbler-4) [reply test/runners/python_spec.js]: Yes, I tested the output using my kumite on CW, but I don't know how it'll be displayed on Qualified.

**Comment by kazk** [comment]
Looks good to me! I'll try to do some testing against existing contents during this week.
Great job on keeping it backwards compatible.
I'm little concerned that the `print("<COMPLETEDIN::>")` workaround some users have been doing will confuse the test output parser though. Hopefully there's not too many and power users can help with finding and removing them.

And thanks for looking into the Unicode issue and `uni_print`!

**Comment by Bubbler-4** [comment]
The workaround should work just fine because, if `describe` or `it` is used old-style, it does not emit `<COMPLETEDIN::>` at all:



Glad you like it :smile:

**Comment by kazk** [comment]
@Bubbler-4 Sorry for the delay. I didn't have the time to test this last week.
Just a quick update. I've deployed this to the new runner environment that I've been testing against mirrored CW traffic and haven't found compatibility issues so far :+1:

**Comment by kazk** [comment]
I found the root cause of the Unicode issue: https://bugs.python.org/issue19846
The image we currently use on Codewars has no locale set (`LC_CTYPE="POSIX"`), forcing ASCII :(

The new runner uses [the official Docker image](https://hub.docker.com/_/python/) with (`LANG C.UTF-8`) and we can just print Unicode without issues.
For Python 2 image, I've also set `PYTHONIOENCODING=UTF-8` because `sys.stdout.encoding` was `None` (setting this will do `.encode("utf-8")` implicitly if I'm understanding correctly).
Overriding the encoding of `stdin`/`stdout`/`stderr` might cause issues for daily use, but should be fine for our use case.

Now we have `print(u'\u2713')` in Python 2.7/3.4/3.6 output `✓` :tada:

**Comment by Voileexperiments** [comment]
When will this be employed to the main CW site (or preview site)? I miss new changes :(

**Comment by kazk** [comment]
I haven't found anything major (looking at the number of successful Python completions) so this should be safe to merge.
@jhoffner what do you think?

If we do, try not to use `test.uni_print`. It's no longer necessary on the new runner so I'll remove that eventually. Current timeline to deploy the new runner for CW is by the end of next week.

**Comment by kazk** [comment]
@Bubbler-4 
preview.codewars.com uses the new runner now if you want to try using this on Kumite. Just be careful not to publish a kata with it.
Thanks for your work!

**Comment by Bubbler-4** [comment]
I just made an [example kumite](https://preview.codewars.com/kumite/5aaa2d33f51c8a313b00000e?sel=5aaa2d33f51c8a313b00000e), and the new framework works perfectly :+1:

Just one thing is, it looks like the runner overhead has increased by ~500ms. It shouldn't be too much of a concern except a few katas that are insanely tight on the bound, but maybe you can consider increasing the time limit to 12500ms.

**Comment by kazk** [comment]
Thanks for trying it out. I realized that `describe`s can be nested too :+1:

The issue with printing Unicode should be resolved as well. I've also avoided concatenation so the error message should be more clear. But I might need to revert it because I noticed some kata uses test methods in preloaded section.

---

Yeah, the wall time (time displayed at the top) won't be the same because the time includes some extra steps. The actual test execution time (time inside test results) shouldn't be that different.

I might make some adjustments later if this is a major problem, but overall the performance should be better and more stable for most of the languages. I'm not too worried because the old runner have inconsistent performance and the wall time can vary more than 500ms.

**Comment by Bubbler-4** [comment]
> But I might need to revert it because I noticed some kata uses test methods in preloaded section.

How about prepending `cw-2.py` import to `setup.py` as well? It's not exposed to the solver, so it doesn't matter if its line numbers are correct or not. Also, that part is the single greatest improvement for kata solvers, so we can't give it up so easily.

**Comment by kazk** [comment]
Yeah, concatenation is terrible for UX. I'll prepend `test = Test = __import__('cw-2')` to `setup.py`.

**Comment by Bubbler-4** [comment]
@kazk 
Blind4Basics made a suggestion to improve `expect_error` so that it can target a specific exception class. [Here](https://preview.codewars.com/kumite/5ab735bee7093b17b2000084?sel=5ab735bee7093b17b2000084) is my kumite that implements the idea, along with some test cases. Is it possible to add this to the new runner?

**Comment by kazk** [comment]
I've added it with slight modification. `expect_error(message, function, exception=Exception)`.

I'll deploy it maybe tomorrow along with other languages that I need to fix.

**Comment by kazk** [comment]
@Bubbler-4 `expect_error` is on preview now, see [this fork](https://preview.codewars.com/kumite/5ab840202619c8f251000029?sel=5ab840202619c8f251000029).

**Comment by kazk** [comment]
I'm going to merge this so @Bubbler-4 gets some credit for it. The actual source that was derived from this will be open sourced later.

**Comment by Blind4Basics** [comment]
Hi guys,

I might have found a way to efficiently forbid the import of modules in python (I doubt it will be bullet proof, I'm not enough in the deep structure of python to know. Though, it seems at least not easy to crack). That could be a nice addition to the test framework, I believe, in the following manner:

* function/data stored in the file of the framework
* a tuple of strings available for the user, each string being the name of a module to forbid (some comments in addition to explain the functioning of it)
* a bit of code to execute just after the preloaded section has been imported and before the import of the file of the user.

I'd send to both of you (@Bubbler-4, @kazk) the code by MP on gitter (I'd rather like that it's not visible for now), so that you can tell me if you think it's worth of the addition.

Cheers,
B4B
